### PR TITLE
also zero pad DHE public key in ClientKeyExchange message for interop

### PR DESCRIPTION
**Pull request by heinzelotto**
The Windows SChannel TLS stack has problems when the DHE pub key is encoded with fewer bytes than the DHE prime.

In https://github.com/openssl/openssl/pull/1320 padding has already been added to work around this problem for the OpenSSL _server_ case, and I experimented a little and Microsoft seems to have also fixed this in their _client_ code.

I guess that OpenSSL server <-> windows client is the more common case, but the opposite case still causes handshakes to randomly fail when a DHE suite is selected. By adding padding for the OpenSSL client key exchange I hope to fix that.

This is my first contribution to OpenSSL, I tried to follow the style guide and conventions. Any corrections are much appreciated.

### CODE REVIEW

File: ssl/statem/statem_clnt.c
Comment (davidben) [inline ssl/statem/statem_clnt.c:3115]: You may want to look at `BN_bn2binpad`.

Comment (heinzelotto) [reply ssl/statem/statem_clnt.c:3115]: Thank you, good idea! I need to close and reopen the pull request because of cla-check.

File: ssl/statem/statem_clnt.c
Comment (davidben) [inline ssl/statem/statem_clnt.c:3095]: `BN_num_bytes(DH_get0_p(dh_clnt))` is probably shorter.

File: ssl/statem/statem_clnt.c
Comment (davidben) [inline ssl/statem/statem_clnt.c:3108]: Likewise, `BN_bn2binpad(DH_get0_pub_key(dh_clnt, ...))`

**Comment by heinzelotto** [comment]
@mattcaswell are you familiar with the WPACKET API and would you taking a look?

File: ssl/statem/statem_clnt.c
Comment (kaduk) [inline ssl/statem/statem_clnt.c:3096]: If we have to make other changes, we could use a capital 'F' to start the comment and a period at the end, to make it a complete sentence, but it's not worth changing just to fix that.
(And the ServerKeyExchange case doesn't do either, anyway.)

File: ssl/statem/statem_clnt.c
Comment (kaduk) [inline ssl/statem/statem_clnt.c:3062]: I'd consider a different name for this variable, maybe `prime_len` -- I'm not actually sure what "np" is supposed to indicate or stand for.

**Comment by kaduk** [comment]
I'm going to dismiss @davidben 's review, since it was before the force-push.
(In some sense it doesn't matter, since we're both committers and this needs an OTC review in order to get merged.)

**Comment by kaduk** [comment]
(the travis failures are the spurious "../../_srcdir/config: no such file or directory" that we've been seeing lately)

**Comment by heinzelotto** [comment]
Thank you @kaduk for your comments. I wanted to ask whether there were any necessary steps on my side once marked as "approval: otc review pending"? And I figured that if I am going to do an additional round trip anyway, I will just add the changes you suggested. (Hopefully I did the fixup! correctly)

**Comment by kaduk** [comment]
@heinzelotto the fixup looks good, thank you.
With the "approval: otc review pending" set we are not waiting on anything else from you (unless that review produces additional feedback/requested changes).  Unfortunately, OTC time is in high demand right now, with the work towards the 3.0 release, so I don't have a great estimate for you.

File: ssl/statem/statem_clnt.c
Comment (t8m) [inline ssl/statem/statem_clnt.c:3104]: Please add check for the return value. Although it is in practice "can't happen" condition, if prime_len is lower than the length of the actual pub_key, the function will fail. You can move the call to the if condition above.

**Review by t8m — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged to master. Thank you for your contribution and patience when waiting for the reviews.

**Comment by paveljanik** [comment]
This makes openssl as the client send out zero padded DHE pubkey in TLSv1.2 which means it breaks the standard as leading zeros should be stripped in TLSv1.2 (as opposite to TLSv1.3 where they should be left in the stream!). Is this really what you need/want?

**Comment by davidben** [comment]
The spec says the resulting secret has leading zeros removed. I believe RFC5246 is silent as to whether the public key is padded. (Which is a mistake. It should have picked one, and the correct one to pick is fixed-width. But a bit late to fix that now, so implementations have to pick one.)

**Comment by paveljanik** [comment]
So on the server side, you have to support (try...) both because your clients could have picked padding and not-padding? ;-)

**Comment by davidben** [comment]
No, just have your parser tolerate leading zeros, as OpenSSL does. The resulting secret is different because it's a KDF input.

(Alternatively, stop doing TLS 1.2 DHE as it's beyond saving.)

**Pull request closed without merging**
